File: python/THWAP/core/threadPools.py
from threading import Thread
import thread, time

# Class: thSpawnThread:
# Authro: Mike "Fuzzy" Partin
# Notes: Spawn a quick thread to do a quick job
class thSpawnThread:
	def __init__(self, func, *args):
		thread.start_new_thread(func, (args))
		return None

# Class: thThreadObj
# Author: Mike "Fuzzy" Partin
# Notes: Generic thread object
class thThreadObj(Thread):
	def __init__(self):
		Thread.__init__(self)
		self.function = None
		self.args = None

	def regWorkLoad(self, function, *args):
		if self.function == None: self.function = function
		if self.args == None: self.args = args

	def run(self):
		if self.args != None:
			self.function(self.args)
		else:
			self.function()

# Class: thThreadPool
# Author: Mike "Fuzzy" Partin
# Notes: Just a pool of generic thread objects
#        Also of note, is this class is not finished
class thThreadPool:
		def __init__(self, threads=5):
			self.pool = []
			for i in range(threads):
					self.pool.append(thThreadObj())
			return None
			obj = thThreadObj()
			obj.regWorkLoad(refillPool, threads)
			obj.run()
			return None

		def refillPool(self, threads):
			while True:
				while len(self.pool) < threads:
					self.pool.append(thThreadObj())
				time.sleep(5)

File: ruby/thwap/core/threadPools.rb
module THWAP
	module Core
		# Class: ThSpawnThread
		# Author: Mike "Fuzzy" Partin
		# Notes: Spawn a quick worker thread.
		# TODO:  Still don't know how to pass the arguments I want,
		class ThSpawnThread
			def initialize(func, *args)
				t = Thread.new{func4()}
				t.join
			end
		end

		class ThThreadObj
			def initialize
				@function = nil
				@args = ()
			end
		
			def regWorkLoad(func, args)
				if @function == nil
					@function = func
				end
				if @args == ()
					@args.push(args)
				end
			end

			def run
				ThSpawnThread.new(@function, @args)
			end
		end

		# Class: ThThreadPool
		# Author: Mike "Fuzzy" Partin
		# Notes: Not done yet, but will be a pool of generic threads
		class ThThreadPool
			def initialize(threads)
				if not threads
					threads = 5
				end
				@threadPool = Array.new 
			  if threads.is_a?(Integer)
					threads.downto(1) { |iter|
						@threadPool.push(ThThreadObj.new)
					}
				end
			end

			def getThreads
				return @threadPool
			end

			def threadCount
				return @threadPool.length
			end

			def regWorkLoad(callback, args)
				return nil
			end
		end
	end
end
